File: LeetCode/DataStruct/Python/Leetcode662-maximum-width-of-binary-tree.py
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        mx = 0
        q = [[root,1]]
        while q:
            n,r,l = len(q),0,q[0][1]
            for i in range(n):
                p = q[0][1] - l + 1
                r,node = q[0][1],q[0][0]
                q = q[1:]
                if node.left:
                    q.append([node.left,p * 2])
                if node.right:
                    q.append([node.right,p * 2 + 1])
            mx = max(mx,r - l + 1)
        return mx

File: LeetCode/DataStruct/JS/Leetcode662-maximum-width-of-binary-tree.js
var widthOfBinaryTree = function(root) {
    let mx = 0;
    let q = [[root,1]];
    while(q.length) {
        const [l,n] = [q[0][1],q.length];
        let r = 0;
        for(let i = 0;i < n;i++) {
            let [node,p] = [q[0][0],q[0][1] - l + 1];
            r = q[0][1];
            q.shift();
            if(node.left != null) {
                q.push([node.left,p * 2]);
            }
            if(node.right != null) {
                q.push([node.right,p * 2 + 1]);
            }
        }
        mx = Math.max(mx,r - l + 1);
    }
    return mx;
};
